### PR TITLE
Samsung duplication bug hack-fix

### PR DESCRIPTION
**Pull request by GaryQian**
This fixes https://github.com/flutter/flutter/issues/31512 barring one edge case which this does not seem to cover.

The hack tells the samsung keyboard that the selection region is -1. This update also passes the composing region bounds, which we are able to set to -1. This apparently tricks the keyboard into resetting their internally cached composing region, which prevents the duplication. The selection region is then immediately restored in the updateEditingState call, so the user never sees the invalid selection.

In addition, I have gated this hack to only happen on Samsung devices in order to reduce the number of strange IMM calls that are being applied on non-relevant keyboards. This is unlikely to cause a regression on non-samsung keyboards, but the gating further reduces the chances.

Tests coming soon.

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 1]
```diff
@@ -132,19 +140,56 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
...
-      // Samsung keyboards to prevent punctuation duplication.
-      CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
-      builder.setComposingText(-1, "");
-      CursorAnchorInfo anchorInfo = builder.build();
-      mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+    // Apply Samsung hacks. Samsung caches composing region data strangely, causing text duplication.
+    if (mIsSamsung) {
+      if (Build.VERSION.SDK_INT >= 21) {
+        // Update the keyboard with a reset/empty composing region. Critical on
+        // Samsung keyboards to prevent punctuation duplication.
```
Comment (cbracken) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Maybe lead with "Samsung keyboards don't clear the composing region on `finishComposingText`. [... current text ...]

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 2]
```diff
@@ -132,19 +140,56 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
...
+  @SuppressLint("NewApi") // New API guard is inline, the linter can't see it.
+  @SuppressWarnings("deprecation")
+  private boolean isSamsung() {
+    InputMethodSubtype subtype = mImm.getCurrentInputMethodSubtype();
+    // Impacted devices all shipped with Android Lollipop or newer.
+    if (subtype == null
+        || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP
+        || !Build.MANUFACTURER.equals("samsung")) {
+      return false;
+    }
```
Comment (cbracken) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: I assume if they fix this bug in a future release, we'll need to change this logic. Is there no way we can narrow this down further?

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 3]
```diff
@@ -132,19 +140,56 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
...
+    String keyboardName =
+        Settings.Secure.getString(
+            mFlutterView.getContext().getContentResolver(), Settings.Secure.DEFAULT_INPUT_METHOD);
+    // The Samsung keyboard is called "com.sec.android.inputmethod/.SamsungKeypad" but look
+    // for "Samsung" just in case Samsung changes the name of the keyboard.
```
Comment (cbracken) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Do they have a history of changing the name of the keyboard and retaining existing bugs? If not, I'd be as precise as possible so we don't over-apply the hacks.

Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: I don't think they change the name of the keyboard, but the bug does manifest in different ways across different devices/versions of the keyboard, sometimes in a regression-like manner (newer versions can have bugs fixed in older versions).

Comment (cbracken) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Makes sense. Thanks.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java (hunk 2 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Overall this hack should be fully compatible with other keyboards. Anything we do here is not 1: telling the keyboard something that it should already know, or 2: performing an action that we are not guaranteed to immediately revert.

It seems unlikely to me that Samsung will release a fix for all versions as well as distribute to the hundreds of millions of devices in the wild.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 4]
```diff
@@ -30,6 +33,9 @@
   private InputMethodManager mImm;
   private final Layout mLayout;
 
+  // Used to determine if Samsung-specific hacks should be applied.
+  private final boolean mIsSamsung;
```
Comment (mklim) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: nit: We actually don't want variables prefixed with `m`, it goes against Google Java style. The other members in the class already have it but I'd avoid matching it going forward. Ideally somebody should come back here later and clean the other names up so they fit the style guide.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 5]
```diff
@@ -132,19 +140,59 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
...
-      mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+    // Apply Samsung hacks. Samsung caches composing region data strangely, causing text
+    // duplication.
+    if (mIsSamsung) {
+      if (Build.VERSION.SDK_INT >= 21) {
+        // Samsung keyboards don't clear the composing region on finishComposingText.
+        // Update the keyboard with a reset/empty composing region. Critical on
+        // Samsung keyboards to prevent punctuation duplication.
+        CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
+        builder.setComposingText(-1, "");
```
Comment (mklim) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: optional/subjective nit here and throughout: It makes API calls like this easier to read if the primitives are annotated so you can tell what they're supposed to be controlling.

```suggestion
        builder.setComposingText(/*composingTextStart=*/-1, /*composingText=*/"");
```

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java [hunk 6]
```diff
@@ -132,19 +140,59 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
...
+        CursorAnchorInfo anchorInfo = builder.build();
+        mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+      }
+      // TODO(garyq): There is still a duplication case that comes from hiding+showing the keyboard.
```
Comment (mklim) [inline shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Hiding and showing the keyboard can restart the IMM. May be related?

Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: Don't have a solid answer here, but in my experimentation, restarting IMM did not seem to have any noticeable effect on this behavior.

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java (hunk 5 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java]: good idea, much clearer what this is doing!

**Comment by tvolkert** [comment]
This PR has been implicated as having introduced a regression in https://github.com/flutter/flutter/issues/53086 and https://github.com/flutter/flutter/issues/53052, both of which are marked as TODAY.

I'll send a revert.

**Comment by tvolkert** [comment]
Hmm, can't be cleanly reverted.  @GaryQian please advise.

**Comment by GaryQian** [comment]
I can manually disable the workarounds for now.

**Comment by GaryQian** [comment]
This is one of those situations where disabling the workaround would re-introduce a number of editing bugs on a subset of samsung devices. Hard to say which set of bugs is "better".

**Pull request merged**